### PR TITLE
Enhancement: Generic funcs

### PR DESCRIPTION
**Pull request by tigh-latte**
Any typed function that performed operations on a slice or a map has been moved over to be generic.

Tests updated too.

### CODE REVIEW

File: maputil/maputil.go [hunk 1]
```diff
@@ -2,10 +2,11 @@
 package maputil // import "github.com/teamwork/utils/maputil"
 
 // Reverse the keys and values of a map.
-func Reverse(m map[string]string) map[string]string {
-	n := make(map[string]string)
+func Reverse[T comparable, V comparable](m map[T]V) map[V]T {
```
Comment (ready4god2513) [inline maputil/maputil.go:5]: `Reverse` should probably be renamed.  I'd argue for `Swap` or something similar.

Comment (Tigh-Gherr) [reply maputil/maputil.go:5]: Agreed, when I seen reverse I had no idea what it did until I read the code, and I'm fairly sure that's a bad sign :thinking:

**Comment by ripexz** [comment]
We probably need to tag current master as `v1.0.0` cause we gonna have to do v2 with these changes 😅

**Comment by rafaeljusto** [comment]
Some other generic function ideas retrieved from [GopherconUK 2022 presentation](https://github.com/feyeleanor/GoGenericsTalk) that could be useful:

### Sum

```go
type Numeric interface {
	~int | ~float32
}

type Indexable[T any] interface {
	~[]T
}

func Sum[T Indexable[E], E Numeric](s T) (r E) {
	for _, x := range s {
		r += x
	}
	return
}
```

### Reduce

```go
type Numeric interface {
	~int | ~float32
}

type Reducible[K comparable, E Numeric] map[K]E

func (m Reducible[K, E]) Reduce(f func(E, E) E) (r E) {
	for _, x := range m {
		r = f(r, x)
	}
	return
}
```

### Map

```go
func Map[T any](s []T, f func(T) T) (r []T) {
	for _, v := range s {
		r = append(r, f(v))
	}
	return
}
```

**Comment by JaTochNietDan** [comment]
Generics in Go, can't be 😨

**Comment by Tigh-Gherr** [comment]
Lmao I forgot about this PR, I think it's travis shite is out of date I need to look into it.

BUT @rafaeljusto they're great shouts

**Comment by ripexz** [comment]
I made some progress on https://github.com/Teamwork/utils/tree/fix-generics branch - switched to GHA and upgraded to go 1.19.

However there seems to be a couple of issues:
- `errorutil` tests fail as stack traces now contain absolute path (I guess they didn't in 1.13? I can't download the old version on the M1 to test...) - this can be circumvented with `-trimpath` flag when building/running but this has additional side effects and I'm not sure it's the right approach.
- `goutil` doesn't seem to work with relative paths anymore, maybe some combination of goroot/gopath can get it running but it seems current approach doesn't quite cut it.

We can probably live with errorutil quirks (probably better to update tests to keep the paths rather than rely on the flag), but goutil probably needs a deeper dive. It seems our only use of it is in kommentaar so arguably since we'll need to tag this as v2, we could lock kommentaar down to the old version and drop it in v2 entirely 🤷

**Comment by ripexz** [comment]
Closing in favour of #73

**Pull request closed without merging**